socs/agents/acti_camera/agent.py

```diff
@@ -0,0 +1,222 @@
+import argparse
+import os
+import shutil
+import time
+from pathlib import Path
+
+import requests
+import txaio
+from ocs import ocs_agent, site_config
+from ocs.ocs_twisted import Pacemaker, TimeoutLock
+
+# For logging
+txaio.use_twisted()
+
+
+class ACTiCameraAgent:
+    """Grab screenshots from ACTi cameras.
+
+    Parameters
+    ----------
+    agent : OCSAgent
+        OCSAgent object which forms this Agent
+    camera_addresses : list
+        List of IP addresses (as strings) of cameras.
+    user : str
+        Username of cameras.
+    password : str
+        Password of cameras.
+
+    Attributes
+    ----------
+    agent : OCSAgent
+        OCSAgent object which forms this Agent
+    is_streaming : bool
+        Tracks whether or not the agent is actively issuing requests to grab
+        screenshots from cameras. Setting to false stops sending commands.
+    log : txaio.tx.Logger
+        txaio logger object, created by the OCSAgent
+    """
+
+    def __init__(self, agent, camera_addresses, locations, user, password):
+        self.agent = agent
+        self.is_streaming = False
+        self.log = self.agent.log
+        self.lock = TimeoutLock()
+
+        self.cameras = []
+        for i, (location, address) in enumerate(zip(locations, camera_addresses)):
+            self.cameras.append({'location': location,
+                                 'address': address,
+                                 'connected': True})
+        self.user = user
+        self.password = password
+
+        agg_params = {
+            'frame_length': 10 * 60  # [sec]
+        }
+        self.agent.register_feed('cameras',
+                                 record=True,
+                                 agg_params=agg_params,
+                                 buffer_time=0)
+
+    @ocs_agent.param('test_mode', default=False, type=bool)
+    def acq(self, session, params=None):
+        """acq(test_mode=False)
+
+        **Process** - Grab screenshots from ACTi cameras.
+
+        Parameters
+        ----------
+        test_mode : bool, optional
+            Run the Process loop only once. Meant only for testing.
+            Default is False.
+
+        Notes
+        -----
+        The most recent data collected is stored in session.data in the
+        structure::
+
+            >>> response.session['data']
+            # for each camera
+            {'camera1': {'location': 'location1',
+                         'last_attempt': 1701983575.032506,
+                         'connected': True,
+                         'address': '10.10.10.41'},
+             'camera2': ...
+        """
+        pm = Pacemaker(1 / 60, quantize=False)
+
+        session.set_status('running')
+        self.is_streaming = True
+        count = 0
+        while self.is_streaming:
+            # Use UTC
+            timestamp = time.time()
+            data = {}
+
+            count += 1
+            for i, camera in enumerate(self.cameras):
```

Comment:
I don't really think the `enumerate` is needed here either. There are a few instances of `self.cameras[i]['connected']` but they should just become `camera['connected']` within this loop.